plugins/dell/fu-plugin-dell.c

```diff
@@ -716,71 +709,61 @@ fu_plugin_dell_detect_tpm (FuPlugin *plugin, GError **error)
 	tpm_guid = fwupd_guid_hash_string (tpm_guid_raw);
 	tpm_id = g_strdup_printf ("DELL-%s" G_GUINT64_FORMAT, tpm_guid);
-
-	tpm_guid_raw_alt = g_strdup_printf ("%04x-%s", system_id, tpm_mode_alt);
-	tpm_guid_alt = fwupd_guid_hash_string (tpm_guid_raw_alt);
-	tpm_id_alt = g_strdup_printf ("DELL-%s" G_GUINT64_FORMAT, tpm_guid_alt);
-
-	g_debug ("Creating primary TPM GUID %s and secondary TPM GUID %s",
-		 tpm_guid_raw, tpm_guid_raw_alt);
 	version_str = fu_common_version_from_uint32 (out->fw_version,
 						     FWUPD_VERSION_FORMAT_QUAD);
-
-	/* make it clear that the TPM is a discrete device of the product */
-	pretty_tpm_name = g_strdup_printf ("TPM %s", tpm_mode);
-	pretty_tpm_name_alt = g_strdup_printf ("TPM %s", tpm_mode_alt);
+	fu_device_set_version (data->tpm_device, version_str, FWUPD_VERSION_FORMAT_QUAD);
 
 	/* build Standard device nodes */
-	if (data->tpm_device != NULL) {
-		dev = g_object_ref (data->tpm_device);
-	} else {
-		dev = fu_device_new ();
-		fu_device_set_id (dev, tpm_id);
-		fu_device_set_vendor (dev, "Dell Inc.");
-		fu_device_set_name (dev, pretty_tpm_name);
-		fu_device_set_summary (dev, "Platform TPM device");
-		fu_device_set_version (dev, version_str, FWUPD_VERSION_FORMAT_QUAD);
-		fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_INTERNAL);
-		fu_device_add_icon (dev, "computer");
-	}
-	fu_device_add_instance_id (dev, tpm_guid_raw);
-	fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_REQUIRE_AC);
-	fu_device_set_metadata (dev, FU_DEVICE_METADATA_UEFI_DEVICE_KIND, "dell-tpm-firmware");
+	fu_device_add_instance_id (data->tpm_device, tpm_guid_raw);
+	fu_device_add_flag (data->tpm_device, FWUPD_DEVICE_FLAG_REQUIRE_AC);
+	fu_device_set_metadata (data->tpm_device, FU_DEVICE_METADATA_UEFI_DEVICE_KIND, "dell-tpm-firmware");
 	if ((out->status & TPM_OWN_MASK) == 0 && out->flashes_left > 0) {
 		if (fu_plugin_dell_capsule_supported (plugin)) {
-			fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_UPDATABLE);
-			fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_NEEDS_REBOOT);
+			fu_device_add_flag (data->tpm_device, FWUPD_DEVICE_FLAG_UPDATABLE);
+			fu_device_add_flag (data->tpm_device, FWUPD_DEVICE_FLAG_NEEDS_REBOOT);
 		} else {
-			fu_device_set_update_error (dev,
+			fu_device_set_update_error (data->tpm_device,
 						    "UEFI capsule updates turned off in BIOS setup");
 		}
-		fu_device_set_flashes_left (dev, out->flashes_left);
+		fu_device_set_flashes_left (data->tpm_device, out->flashes_left);
 	} else {
-		fu_device_set_update_error (dev,
+		fu_device_set_update_error (data->tpm_device,
 					    "Updating disabled due to TPM ownership");
 	}
 	/* build GUIDs from TSS strings */
-	if (!fu_plugin_dell_add_tpm_model (dev, &error_tss))
+	if (!fu_plugin_dell_add_tpm_model (data->tpm_device, &error_tss))
 		g_debug ("could not build instances: %s", error_tss->message);
 
-	if (!fu_device_setup (dev, error))
-		return FALSE;
-	if (data->tpm_device == NULL)
-		fu_plugin_device_register (plugin, dev);
```

Comment:
So I think the problem with this flow is that the device is still owned by `tpm` not by `uefi`.  So the `uefi` registration won't pick it up.

Reply:
So maybe I'm making this all too complicated. Perhaps it's better is we just have:

* A tpm plugin, where the `dell` plugin is marked BETTER_THAN the `tpm` plugin, so that if the dell plugin creates the TPM device the tpm-created one gets ignored. Sure, there's code duplication, but much less risk we cause a regression somewhere.

* A tpm-eventlog plugin, which adds the eventlog child functionality to *either* tpm device.

Reply:
> A tpm plugin, where the dell plugin is marked BETTER_THAN the tpm plugin, so that if the dell plugin creates the TPM device the tpm-created one gets ignored. Sure, there's code duplication, but much less risk we cause a regression somewhere.

Yeah that means a lot less complexity of this dance between 3rd plugins then.

> A tpm-eventlog plugin, which adds the eventlog child functionality to either tpm device.

Yeah that makes sense - it can listen for the registered signal and look for something indicating it's a TPM and then mark that as a parent.

Reply:
> it can listen for the registered signal

I don't even think it needs to do that. Other plugins just use `fu_device_add_parent_guid`

Reply:
Oh and make a common GUID for both that is platform-tpm or something. Yeah

Reply:
I went with `system-tpm` -- is that better or worse than `platform-tpm`?

Reply:
That's fine too. No strong opinion